proposals/p2240.md

```diff
@@ -0,0 +1,670 @@
+contains a `[:]` operator but not the enclosing pack operator. In those cases,
+the result of the rewrite is an expression headed by a `[:]` operator, which
+means that its value is a pack. Thus, we can say that
+`([:]("a", "b", "c"), [:](1, 2, 3))` rewrites to
+`[:](("a", 1), ("b", 2), ("c", 3))`.
+
+### Pattern semantics
+
+Pack expansions can also appear in patterns. The semantics are chosen to follow
+the general principle that pattern matching is the inverse of expression
+evaluation, so for example if the pattern `(..., [:]x: String)` matches some
+scrutinee value `s`, the expression `(..., [:]x)` should be equal to `s`.
+
+These are run-time semantics, so the scrutinee expression is fully evaluated
+(doesn't contain pack expansions).
+
+**NOTE TO READER:** This section is specifying run-time semantics, but I'm
+paying special attention to the conditions for pattern irrefutability because
+that's going to be an important property when typechecking variadic patterns.
+
+_Tuple pattern:_
+
+There can be no more than one `...,` pattern in a tuple pattern.
+
+The N elements of the pattern before the `...,` expansion are matched with the
+first N elements of the scrutinee, and the M elements of the pattern after the
+`...,` expansion are matched with the last M elements of the scrutinee. If the
+scrutinee does not have at least N + M elements, the pattern does not match. The
+operand of the `...,` pattern is iteratively matched with each of the remaining
+K scrutinee elements, if any. The iterations are numbered sequentially from 1 to
+K, and the number of the current iteration is called the _pack index_.
+
+Observe that the pattern is irrefutable for a given scrutinee type if:
+
+-   At most one `...,` pattern
+-   Scrutinee type is a tuple type with at least N + M elements
+-   First N and last M pattern elements are irrefutable given the corresponding
+    scrutinee element types
+-   For all I from 1 to K, the `...,` operand pattern with pack index I is
+    irrefutable given the type of the Ith scrutinee element. (Note that the
+    typechecker doesn't know K)
+
+_Identifier that names a deduced parameter_
+
+The deduced parameter is unified with the scrutinee value.
+
+Pattern is irrefutable for a given scrutinee type if:
+
+-   Unification does not reach a contradiction.
+
+The typechecker must also ensure that the deduced value is a valid value of the
+deduced parameter type.
+
+_Name binding pattern:_
+
+If the type subpattern contains a "free" use of the `[:]` operator, the name
+binding pattern behaves at run-time as if it were prefixed with `[:]` (see
+below).
+
+Otherwise, if the pattern is inside a pack expansion, but not inside a pack
+argument pattern, the program is ill-formed (because the name would be bound
+multiple times).
+
+Name binding patterns are always irrefutable at run-time. However, typechecker
+must also ensure the bound value is a valid value of the binding type.
+
+_`[:]` pattern:_
```

Comment:
Example?

Reply:
My issue here is that I don't think "`[:]` pattern" is on the list of kinds of patterns like "tuple pattern" and "binding pattern" are. Without an example of what you mean, I don't know what the subject of this section is.